File: bullet.py
"""World-space projectile entity."""

from math import cos, sin

import pygame

import background as bG
import uiTheme as ui
import variableHolster as vH


class Bullet:
    """A projectile whose path is independent of player and camera movement."""

    def __init__(
            self,
            world_x, world_y,
            speed,
            direction,
            bullet_range,
            size,
            color,
            pierce,
            damage,
            is_critical,
            ):
        self.worldX = world_x
        self.worldY = world_y
        self.posX, self.posY = bG.world_to_screen(world_x, world_y)
        self.speed = speed
        self.direc = direction
        self.size = size
        self.color = color
        self.bRange = bullet_range
        self.bPierce = pierce
        self.remFlag = False
        self.damage = damage
        self.currCrit = is_critical
        self.portalCooldown = 0.0

    def updateAndDrawBullet(self, screen):
        # Advance in world space using only the projectile's own velocity. Camera and
        # player movement affect the final screen projection, never the bullet path.
        distance = self.speed * vH.get_frame_scale()
        seconds = vH.get_timer_step() / max(1, vH.frameRate)
        self.portalCooldown = max(0.0, self.portalCooldown - seconds)
        self.worldX += cos(self.direc) * distance
        self.worldY -= sin(self.direc) * distance
        self.bRange -= distance

        self.posX, self.posY = bG.world_to_screen(self.worldX, self.worldY)
        world_rect = pygame.Rect(self.worldX, self.worldY, self.size, self.size)
        rect = pygame.Rect(self.posX, self.posY, self.size, self.size)
        pygame.draw.rect(screen, ui.INK, rect.inflate(4, 4))
        pygame.draw.rect(screen, ui.PURPLE if self.currCrit else ui.CREAM, rect)
        pygame.draw.rect(screen, ui.TEXT, rect.inflate(-int(self.size * .5), -int(self.size * .5)))

        if bG.rect_hits_wall(world_rect) or self.bRange <= 0:
            self.remFlag = True

File: csharp/RotBoiRemastered/Entities/Bullet.cs
using Microsoft.Xna.Framework;
using Microsoft.Xna.Framework.Graphics;
using RotBoiRemastered.Core;
using RotBoiRemastered.UI;
using RotBoiRemastered.World;

namespace RotBoiRemastered.Entities;

/// <summary>
/// World-space player projectile whose path is independent of player and
/// camera movement. Ported from bullet.py. Update (movement/removal-flag)
/// and Draw are split, unlike Python's combined updateAndDrawBullet, so
/// removal logic is unit testable without a GraphicsDevice.
///
/// `PortalCooldown` is unused by anything ported so far -- it exists for
/// bossTypes.py's projectile-portal interception (not yet ported), which
/// reads/writes it directly. Kept as a public settable field for that
/// future caller rather than re-deriving it later.
/// </summary>
public sealed class Bullet
{
    public float WorldX { get; private set; }
    public float WorldY { get; private set; }
    public float Speed { get; }
    public float Direction { get; }
    public float Size { get; }
    public Color Color { get; }
    public float Range { get; private set; }
    public int Pierce { get; set; }
    public float Damage { get; }
    public bool IsCritical { get; }
    public bool RemFlag { get; private set; }
    public float PortalCooldown { get; set; }

    public Bullet(float worldX, float worldY, float speed, float direction, float bulletRange,
        float size, Color color, int pierce, float damage, bool isCritical)
    {
        WorldX = worldX;
        WorldY = worldY;
        Speed = speed;
        Direction = direction;
        Size = size;
        Color = color;
        Range = bulletRange;
        Pierce = pierce;
        Damage = damage;
        IsCritical = isCritical;
    }

    public Rectangle WorldRect() => new((int)WorldX, (int)WorldY, (int)Size, (int)Size);

    public void Update(Battleground battleground)
    {
        float distance = Speed * (float)Simulation.GetFrameScale();
        float seconds = (float)Simulation.GetTimerStep() / Math.Max(1, Simulation.FrameRate);
        PortalCooldown = Math.Max(0f, PortalCooldown - seconds);
        WorldX += MathF.Cos(Direction) * distance;
        WorldY -= MathF.Sin(Direction) * distance;
        Range -= distance;

        if (battleground.RectHitsWall(WorldRect()) || Range <= 0)
            RemFlag = true;
    }

    public void Draw(SpriteBatch spriteBatch, Camera camera, Vector2 playerWorldPosition, Vector2 screenShake)
    {
        Vector2 screenPosition = camera.WorldToScreen(new Vector2(WorldX, WorldY), playerWorldPosition, screenShake);
        var rect = new Rectangle((int)screenPosition.X, (int)screenPosition.Y, (int)Size, (int)Size);
        rect.Inflate(4, 4);
        Primitives2D.FillRect(spriteBatch, rect, UiTheme.Ink);
        rect.Inflate(-4, -4);
        Primitives2D.FillRect(spriteBatch, rect, IsCritical ? UiTheme.Purple : UiTheme.Cream);
        var inner = rect;
        inner.Inflate(-(int)(Size * 0.5f), -(int)(Size * 0.5f));
        Primitives2D.FillRect(spriteBatch, inner, UiTheme.Text);
    }
}
